quality_check/09-prior_information_matrix.Rmd

```diff
@@ -0,0 +1,225 @@
+  dplyr::select(-Gene) %>%
+  # drop records missing ZFIN ids
+  dplyr::filter(!is.na(ZFIN)) %>%
+  # unique records only, just in case
+  dplyr::distinct() %>%
+  # get binary matrix, where the first column will be gene and the columns are
+  # pathways
+  reshape2::dcast(ZFIN ~ Pathway, fun.aggregate = length)
+```
+
+```{r}
+# now a more PLIER friendly format
+zebrafish_wp_binary_mat <- zebrafish_wp_binary_df %>%
+  tibble::column_to_rownames("ZFIN") %>%
+  as.matrix()
+saveRDS(zebrafish_wp_binary_mat, 
+        file.path(pathway_dir, "Danio_rerio_wikipathways_prior_info.RDS"))
+```
+
+## ZFIN
+
+Given the [licensing of the ZFIN resource](https://zfin.org/warranty.html) and 
+the fact that downloading via the web interface has the advantage of including 
+the archive date in the file name/header and including the column names, we've 
+obtained these files via the 
+[downloads webpage](https://zfin.org/downloads/archive/2019.01.25) using the
+date that is currently the most recent archive (`2019-01-25`).
+
+
+```{r}
+wildtype_expression_file <- 
+  file.path(pathway_dir, "wildtype-expression_fish_2019.01.25.txt")
+# we're only going to use the in situ data
+wildtype_expression_df <- readr::read_tsv(wildtype_expression_file, skip = 1) %>%
+  dplyr::filter(Assay == "mRNA in situ hybridization") %>%
+  dplyr::select(-X15)
+```
+
+We're hoping to get two gene sets from this ZFIN information: 1) anatomical
+structure and 2) developmental stage.
+Parsing this by stage-tissue pairs ends up being extremely sparse, so we can't
+approach it that way and get anything useful.
+
+### Stages
+
+```{r}
+stage_file <- file.path(pathway_dir, "stage_ontology_2019.01.25.txt")
+stage_df <- readr::read_tsv(stage_file, skip = 1) %>%
+  # drop column associated with parsing error
+  dplyr::select(-X6)
+```
+
+```{r}
+stage_ordered_factor <- factor(stage_df$`Stage Name`[-1],  # drop unknown
+                               levels = stage_df$`Stage Name`[-1],
+                               ordered = TRUE)
+stage_ordered_factor
+```
+```{r}
+gene_stages_list <- list()
+for (current_gene in unique(wildtype_expression_df$`Gene ID`)) {
+  
+  # initialize vector to hold _all_ the stages for the current gene
+  current_gene_stages <- vector()
+  
+  # data.frame filtered to current gene only
+  current_gene_df <- wildtype_expression_df %>%
+    dplyr::filter(`Gene ID` == current_gene)
+  
+  # all rows in current_gene_df
+  for (gene_row_iter in 1:nrow(current_gene_df)) {
+    # pull pertinent stage information from this row
+    start_stage <- current_gene_df$`Start Stage`[gene_row_iter]
+    end_stage <- current_gene_df$`End Stage`[gene_row_iter]
+    
+    # find the indices corresponding to the stages and extract the stage
+    # names from the ordered factor vector
+    start_index <- which(stage_ordered_factor == start_stage)
+    end_index <- which(stage_ordered_factor == end_stage)
+    covered_stages <- as.character(stage_ordered_factor[start_index:end_index])
```

Comment:
I assume with this set up there's no risk of these indices being out of order or having gaps in between the rows you need?

Reply:
You are correct. I added a check for `end_index >= start_index` just in case and nothing fails that check. Gaps in a start-end range like this shouldn't be possible.